lib/puppet/http/service/compiler.rb

```diff
@@ -129,6 +129,70 @@ def post_catalog(name, facts:, environment:, configured_environment: nil, transa
     [response, deserialize(response, Puppet::Resource::Catalog)]
   end
 
+  # TODO; document this API once it is finalized
+  def post_catalog4(certname, persistence, environment, facts: nil, trusted_facts: nil, transaction_uuid: nil, job_id: nil, options: nil)
+    body = {
+      certname: certname,
+      persistence: persistence,
+      environment: environment,
+      transaction_uuid: transaction_uuid,
+      job_id: job_id,
+      options: options
+    }
+    body.merge!({ facts: { values: facts }}) unless facts.nil?
+    body.merge!({ trusted_facts: { values: trusted_facts }}) unless trusted_facts.nil?
+    headers = {
+      'accept-encoding' => Puppet::Network::HTTP::Compression::ACCEPT_ENCODING,
+      'Content-Type' => 'text/json',
+      'Accept' => 'application/json',
+      Puppet::Network::HTTP::HEADER_PUPPET_VERSION => Puppet.version
```

Comment:
The client will handle `accept-encoding` and `HEADER_PUPPET_VERSION` automatically, so you can simplify this to:

```ruby
    headers = add_puppet_headers(
      'Accept' => get_mime_types(Puppet::Resource::Catalog).join(', '),
      'Content-Type' => 'application/json'
    )
```